### PR TITLE
Support hidden parameters in global_parameter

### PR DESCRIPTION
**Pull request by mdellweg**
affects #1024

### CODE REVIEW

File: plugins/module_utils/foreman_helper.py [hunk 1]
```diff
@@ -511,6 +511,24 @@ def _patch_cv_filter_rule_api(self):
             if create_param is not None and update_param is None:
                 _content_view_filter_rule_update['params'].append(create_param)
 
+    @_check_patch_needed()
+    def _patch_common_parameter_index_show_api(self):
+        """
+        This is a workaround for wrong validation of the 'show_missing' querystring in 'common_parameter' index and show.
+        """
```
Comment (mdellweg) [inline plugins/module_utils/foreman_helper.py]: Not sure if this is an api bug or an apidoc bug.

File: plugins/modules/global_parameter.py [hunk 2]
```diff
@@ -111,14 +111,18 @@
 
 
 class ForemanCommonParameterModule(ForemanEntityAnsibleModule):
-    pass
+    def remove_sensitive_fields(self, entity):
+        if entity and 'hidden_value?' in entity:
+            entity['hidden_value'] = entity.pop('hidden_value?')
+        return super(ForemanCommonParameterModule, self).remove_sensitive_fields(entity)
```
Comment (mdellweg) [inline plugins/modules/global_parameter.py:126]: Should we remove value if `hidden_value` is `True`?

File: plugins/module_utils/foreman_helper.py (hunk 1 above)
Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:518]: You mean show_hidden, not show_missing, right? Why is bool not the right validator here?

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py]: You mean show_hidden, not show_missing, right? Why is bool not the right validator here?

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py]: Or do we just fail to encode them properly when they are passed as GET params and not inside JSON?

File: plugins/modules/global_parameter.py (hunk 2 above)
Comment (evgeni) [reply plugins/modules/global_parameter.py:126]: I think that's reasonable, yes

**Review by evgeni — COMMENTED** [review]
Comments

File: plugins/module_utils/foreman_helper.py (hunk 1 above)
Comment (mdellweg) [reply plugins/module_utils/foreman_helper.py]: Yes, show_hidden.

apidoc:
https://theforeman.org/api/2.1/apidoc/v2/common_parameters/index.html
api:
https://github.com/theforeman/foreman/blob/ae96cb70130968bdcc8b7a9a9af7edc7c6e0d9a8/app/views/api/v2/common/show_hidden.json.rabl

It only works with string `"true"`.

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py]: Right. GET params are always strings, as that's the only thing that you can pass (contrary to JSON which has types). The API doc defines it as bool, so the api could potentially convert it properly, but doesn't.

I think the right fix here is not to patch the apidoc but make our code (or rather, apypie) send booleans as true/false (lowercase strings) when passing them as GET params (right now it sends True/False (well, `str(value)`)).

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py]: FWIW, the api should also accept the string "1", which it currently doesn't, but that's not our problem.

Comment (mdellweg) [reply plugins/module_utils/foreman_helper.py]: so it's an apypie bug?

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py]: I think it is, yeah.

File: plugins/modules/global_parameter.py [hunk 3]
```diff
@@ -132,7 +145,7 @@ def main():
     )
 
     with module.api_connection():
-        entity = module.lookup_entity('entity')
+        entity = module.lookup_entity('entity', params={'show_hidden': True})
```
Comment (evgeni) [inline plugins/modules/global_parameter.py:148]: smart_class_parameter also has support for `hidden_value`, but doesn't pass `show_hidden`, is that a bug (in smart_class_parameter)?

File: plugins/modules/global_parameter.py [hunk 4]
```diff
@@ -141,6 +154,8 @@ def main():
             module.foreman_params['value'] = parameter_value_to_str(module.foreman_params['value'], module.foreman_params['parameter_type'])
             if entity and 'value' in entity:
                 entity['value'] = parameter_value_to_str(entity['value'], entity.get('parameter_type', 'string'))
+            if entity and 'hidden_value?' in entity:
+                entity['hidden_value'] = entity.pop('hidden_value?')
```
Comment (evgeni) [inline plugins/modules/global_parameter.py:158]: this smells like we'll have a HiddenValueEntity at some point, as the same code also exists in smart_class_parameter ;)

File: changelogs/fragments/global_parameter-hidden_values.yaml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - allow to set hidden flag on ``global_parameter`` (https://github.com/theforeman/foreman-ansible-modules/issues/1024)
```
Comment (evgeni) [inline changelogs/fragments/global_parameter-hidden_values.yaml]:
```suggestion
  - global_parameter - allow to set hidden flag (https://github.com/theforeman/foreman-ansible-modules/issues/1024)
```

**Comment by evgeni** [comment]
the host and activation_key fixtures look broken, but I don't think you had to re-record them anyways?

File: tests/test_playbooks/fixtures/activation_key-11.yml [hunk 6]
```diff
@@ -314,12 +252,14 @@ interactions:
...
-      Cookie:
-      - _session_id=4981a4276e1accb5572d962fd7ebc583
       User-Agent:
       - apypie (https://github.com/Apipie/apypie)
     method: GET
+<<<<<<< HEAD
     uri: https://centos7-katello-3-14.yatsu.example.com/katello/api/activation_keys/3/product_content?content_access_mode_all=true
+=======
+    uri: https://foreman.example.org/katello/api/activation_keys/3/product_content?content_access_mode_all=true
+>>>>>>> 128f9666... fixup! Support hidden parameters in global_parameter
```
Comment (evgeni) [inline tests/test_playbooks/fixtures/activation_key-11.yml]: yeah, broken

**Comment by mdellweg** [comment]
> the host and activation_key fixtures look broken, but I don't think you had to re-record them anyways?

Technically i had to, before you "fixed" them in #1013

File: plugins/modules/global_parameter.py (hunk 3 above)
Comment (mdellweg) [reply plugins/modules/global_parameter.py:148]: That probably means that we do not yet support hidden values for smart_class_parameter.
We are also still missing that feature for nested parameter in host, domain, you-name-it.

File: plugins/modules/global_parameter.py (hunk 4 above)
Comment (mdellweg) [reply plugins/modules/global_parameter.py:158]: :nose: not today?

Comment (evgeni) [reply plugins/modules/global_parameter.py:158]: absolutely, just wanted to raise awareness

File: plugins/modules/global_parameter.py (hunk 3 above)
Comment (evgeni) [reply plugins/modules/global_parameter.py:148]: not support them in an idempotent way? but yeah.

**Review by evgeni — APPROVED** [review]

**Pull request merged**